packages/core/src/app/pg-core/db/chain-db.ts

```diff
@@ -1,24 +1,523 @@
+    )
+
+    // We can't create a parent if either child is node.
+    if (leftChild === null || rightChild === null) {
+      return null
+    }
+
+    // We have both children so we can compute the parent!
+    return computeParent(leftChild, rightChild)
+  }
+
+  /**
+   * Adds a Merkle tree node for a specific block. Won't add the node if it
+   * already exists in the database or can be computed from other nodes that
+   * exist.
+   * @param blockNumber Block to add the node for.
+   * @param nodeIndex Index of the node to add.
+   * @param nodeHash Hash of the node to add.
+   */
+  public async addMerkleTreeNode(
+    blockNumber: number,
+    nodeIndex: number,
+    nodeHash: string
+  ): Promise<void> {
+    // Don't do anything if we already have the node.
+    if ((await this.getMerkleTreeNode(blockNumber, nodeIndex)) !== null) {
+      return
+    }
+
+    // Smart parent checks. We don't need to store the parent as long as we
+    // have this node's sibling. Parent can simply be re-generated later on.
+    // Skip if this is the root node because it doesn't have a sibling.
+    if (nodeIndex !== 0) {
+      const siblingIndex = getSiblingIndex(nodeIndex)
+      if ((await this.getMerkleTreeNode(blockNumber, siblingIndex)) !== null) {
+        const parentIndex = getParentIndex(nodeIndex)
+        await this.removeMerkleTreeNode(blockNumber, parentIndex)
+      }
+    }
+
+    // We don't have the node, add it.
+    const key = KEYS.TREE_NODES.encode([blockNumber, nodeIndex])
+    const value = Buffer.from(nodeHash)
+    await this.db.put(key, value)
+  }
+
+  /**
+   * Recursively deletes Merkle tree nodes. Intelligently deletes children
+   * that could be used to re-generate the node.
+   * @param blockNumber Block number to delete nodes from.
+   * @param nodeIndex Index of the node to delete.
+   */
+  public async removeMerkleTreeNode(
+    blockNumber: number,
+    nodeIndex: number
+  ): Promise<void> {
+    // We don't have the node and we can't generate it from children.
+    const node = await this.getMerkleTreeNode(blockNumber, nodeIndex)
+    if (node === null) {
+      return
+    }
+
+    // We have the node value, but we don't know whether we actually have the
+    // value stored or if we have children that can re-generate the node.
+
+    // First, check if we have the node in question.
+    const key = KEYS.TREE_NODES.encode([blockNumber, nodeIndex])
+    const value = await this.db.get(key)
+
+    // We have this specific node, so operate on it.
+    if (value !== null) {
+      // Delete the node.
+      await this.db.del(key)
+
+      // Figure out if this node had a sibling. If so, we're going to want to
+      // compute the parent and insert it since we'd be losing that info.
+      const siblingIndex = getSiblingIndex(nodeIndex)
+      const sibling = await this.getMerkleTreeNode(blockNumber, siblingIndex)
+
+      // We have a sibling, so we're going to compute the parent and insert it.
```

Comment:
We had a sibling, meaning we could've generated the parent. If we simply deleted this node, then we'd lose that parent information. We therefore want to insert the parent again so we don't lose any information other than what we specifically wanted to remove.